File: src/main/java/com/github/pareronia/i18n_puzzles/Puzzle2025_09.java
package com.github.pareronia.i18n_puzzles;

import static java.util.stream.Collectors.joining;

import java.time.LocalDate;
import java.time.Month;
import java.time.format.DateTimeFormatter;
import java.time.format.DateTimeParseException;
import java.time.format.ResolverStyle;
import java.util.Arrays;
import java.util.HashMap;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.Set;

import com.github.pareronia.i18n_puzzles.common.Sample;
import com.github.pareronia.i18n_puzzles.common.Samples;
import com.github.pareronia.i18n_puzzles.common.SolutionBase;

public class Puzzle2025_09 extends SolutionBase<String>{

    private final LocalDate NINE_ELEVEN = LocalDate.of(2001, Month.SEPTEMBER, 11);
    
    private Puzzle2025_09(final boolean debug) {
        super(debug);
    }

    public static Puzzle2025_09 create() {
        return new Puzzle2025_09(false);
    }

    public static Puzzle2025_09 createDebug() {
        return new Puzzle2025_09(true);
    }

    private boolean valid(final String date, final DateFormat df) {
        try {
            df.get().parse(date);
            return true;
        } catch (final DateTimeParseException e) {
            return false;
        }
    }

    @Override
    public String solve(final List<String> input) {
        record Entries(String date, String[] names) {
            public static Entries fromInput(final String line) {
                final String[] splits = line.split(": ");
                return new Entries(splits[0], splits[1].split(", "));
            }
        }
        
        final List<Entries> entrieses = input.stream()
                .map(Entries::fromInput)
                .toList();
        final Map<String, Integer> map = new HashMap<>();
        for (final Entries e : entrieses) {
            final int matches = Arrays.stream(DateFormat.values())
                .filter(df -> valid(e.date, df))
                .mapToInt(df -> df.mask)
                .reduce(0, (a, b) -> a | b);
            for (final String name : e.names) {
                map.merge(name, matches, (a, b) -> a & b);
            }
        }
        final Set<String> ans = new HashSet<>();
        for (final Entries e : entrieses) {
            for (final String name : e.names) {
                for (final DateFormat df : DateFormat.values()) {
                    if (((df.mask & map.get(name)) != 0)
                            && e.date.equals(df.get().format(NINE_ELEVEN))) {
                        ans.add(name);
                    }
                }
            }
        }
        return ans.stream().sorted().collect(joining(" "));
    }

    @Override
    @Samples({
        @Sample(input = TEST, expected = "Margot Peter")
    })
    public void samples() { }

    public static void main(final String[] args) throws Exception {
        Puzzle2025_09.create().run();
    }

    private static final String TEST = """
            16-05-18: Margot, Frank
            02-17-04: Peter, Elise
            06-02-29: Peter, Margot
            31-09-11: Elise, Frank
            09-11-01: Peter, Frank, Elise
            11-09-01: Margot, Frank
            """;

    private enum DateFormat {
        DMY("dd-MM-uu", 0b0001),
        MDY("MM-dd-uu", 0b0010),
        YMD("uu-MM-dd", 0b0100),
        YDM("uu-dd-MM", 0b1000);
        
        DateFormat(final String format, final int mask) {
            this.value = format;
            this.mask = mask;
        }

        private final String value;
        private final int mask;

        public DateTimeFormatter get() {
            return DateTimeFormatter
                    .ofPattern(this.value)
                    .withResolverStyle(ResolverStyle.STRICT);
        }
    }
}

File: src/main/python/Puzzle2025_09.py
from __future__ import annotations

import itertools
import operator
import sys
from datetime import datetime
from enum import Enum
from enum import unique
from functools import reduce

from i18n_puzzles.common import InputData
from i18n_puzzles.common import SolutionBase
from i18n_puzzles.common import i18n_puzzles_samples

TEST = """\
16-05-18: Margot, Frank
02-17-04: Peter, Elise
06-02-29: Peter, Margot
31-09-11: Elise, Frank
09-11-01: Peter, Frank, Elise
11-09-01: Margot, Frank
"""

Output = str
NINE_ELEVEN = datetime(year=2001, month=9, day=11)


@unique
class DateFormat(Enum):
    format: str
    mask: int

    def __new__(cls, format: str, mask: int) -> DateFormat:
        obj = object.__new__(cls)
        obj._value_ = format
        obj.mask = mask
        return obj

    DMY = ("%d-%m-%y", 0b0001)
    MDY = ("%m-%d-%y", 0b0010)
    YMD = ("%y-%m-%d", 0b0100)
    YDM = ("%y-%d-%m", 0b1000)

    def is_valid(self, date: str) -> bool:
        try:
            datetime.strptime(date, self.value)
            return True
        except ValueError:
            return False


class Solution(SolutionBase[Output]):
    def solve(self, input: InputData) -> Output:
        def parse(line: str) -> tuple[str, list[str]]:
            date, names = line.split(": ")
            return date, names.split(", ")

        entries = [parse(line) for line in input]
        d = dict[str, int]()
        for date, names in entries:
            matches = reduce(
                operator.or_,
                (df.mask for df in DateFormat if df.is_valid(date)),
            )
            for name in names:
                d[name] = d[name] & matches if d.get(name) else matches
        return " ".join(
            sorted(
                name
                for date, names in entries
                for name, df in itertools.product(names, DateFormat)
                if (
                    df.mask & d[name]
                    and NINE_ELEVEN.strftime(df.value) == date
                )
            )
        )

    @i18n_puzzles_samples(((TEST, "Margot Peter"),))
    def samples(self) -> None:
        pass


solution = Solution(9)


def main() -> None:
    solution.run(sys.argv)


if __name__ == "__main__":
    main()
